targets/TARGET_STM/can_api.c

```diff
@@ -17,6 +17,496 @@
+
+    return can_internal_init(obj);
+}
+
+
+/** Filter out incoming messages
+ *
+ *  @param obj CAN object
+ *  @param id the id to filter on
+ *  @param mask the mask applied to the id
+ *  @param format format to filter on
+ *  @param handle message filter handle (not supported yet)
+ *
+ *  @returns
+ *    0 if filter change failed or unsupported,
+ *    new filter handle if successful (not supported yet => returns 1)
+ */
+int can_filter(can_t *obj, uint32_t id, uint32_t mask, CANFormat format, int32_t handle)
+{
+    UNUSED(handle); // Not supported yet (seems to be a used in read function?)
+
+    FDCAN_FilterTypeDef sFilterConfig = {0};
+
+    if (format == CANStandard) {
+        sFilterConfig.IdType = FDCAN_STANDARD_ID;
+        sFilterConfig.FilterIndex = 0;
+        sFilterConfig.FilterType = FDCAN_FILTER_MASK;
+        sFilterConfig.FilterConfig = FDCAN_FILTER_TO_RXFIFO0;
+        sFilterConfig.FilterID1 = id;
+        sFilterConfig.FilterID2 = mask;
+    } else if (format == CANExtended) {
+        sFilterConfig.IdType = FDCAN_EXTENDED_ID;
+        sFilterConfig.FilterIndex = 0;
+        sFilterConfig.FilterType = FDCAN_FILTER_MASK;
+        sFilterConfig.FilterConfig = FDCAN_FILTER_TO_RXFIFO0;
+        sFilterConfig.FilterID1 = id;
+        sFilterConfig.FilterID2 = mask;
+    } else { // Filter for CANAny format cannot be configured for STM32
+        return 0;
+    }
+
+    if (HAL_FDCAN_ConfigFilter(&obj->CanHandle, &sFilterConfig) != HAL_OK) {
+        return 0;
+    }
+
+    return 1;
+}
+
+
+int can_write(can_t *obj, CAN_Message msg, int cc)
+{
+    FDCAN_TxHeaderTypeDef TxHeader = {0};
+
+    UNUSED(cc);
+
+    // Configure Tx buffer message
+    TxHeader.Identifier = msg.id;
+    if (msg.format == CANStandard) {
+        TxHeader.IdType = FDCAN_STANDARD_ID;
+    } else {
+        TxHeader.IdType = FDCAN_EXTENDED_ID;
+    }
+
+    TxHeader.TxFrameType = FDCAN_DATA_FRAME;
+    TxHeader.DataLength = msg.len << 16;
+    TxHeader.ErrorStateIndicator = FDCAN_ESI_ACTIVE;
+    TxHeader.BitRateSwitch = FDCAN_BRS_OFF;
+    TxHeader.FDFormat = FDCAN_CLASSIC_CAN;
+    TxHeader.TxEventFifoControl = FDCAN_STORE_TX_EVENTS;
+    TxHeader.MessageMarker = 0;
+
+    if (HAL_FDCAN_AddMessageToTxFifoQ(&obj->CanHandle, &TxHeader, msg.data) != HAL_OK) {
+        // printf("*** can_write error %x\n", HAL_FDCAN_GetError(&obj->CanHandle));
+        return 0;
+    }
+
+    return 1;
+}
+
+int can_read(can_t *obj, CAN_Message *msg, int handle)
+{
+    UNUSED(handle); // Not supported yet (seems to be a handle to a filter configuration?)
+
+    if (HAL_FDCAN_GetRxFifoFillLevel(&obj->CanHandle, FDCAN_RX_FIFO0) == 0) {
+        return 0; // No message arrived
+    }
+
+    FDCAN_RxHeaderTypeDef RxHeader = {0};
+    if (HAL_FDCAN_GetRxMessage(&obj->CanHandle, FDCAN_RX_FIFO0, &RxHeader, msg->data) != HAL_OK) {
+        error("HAL_FDCAN_GetRxMessage error\n");
```

Comment:
Should it used mbed_error?

Reply:
HAL_FDCAN_GetRxMessage should not return 0 after that HAL_FDCAN_GetRxFifoFillLevel has returned something.

Reply:
I am just wondering if we should use `mbed_error` that can contain more information rather than older `error` function (will print unknown error). 

@SenRamakri Can you review this, we still use error in targets implementation?